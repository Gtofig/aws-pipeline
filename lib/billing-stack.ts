import { Stack, StackProps } from "aws-cdk-lib";
import { Budget } from "./constructs/budget";
import { Construct } from "constructs";

interface BillingStackProps extends StackProps {
  budgetAmount: number;
  emailAddress: string;
}

export class BillingStack extends Stack {
  constructor(scope: Construct, id: string, props: BillingStackProps) {
    super(scope, id, props);

    new Budget(this, "Budget", {
      budgetAmount: props.budgetAmount,
      emailAddress: props.emailAddress,
    });
  }
}
