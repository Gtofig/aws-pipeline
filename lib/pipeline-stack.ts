import * as cdk from "@aws-cdk/core";
import { Artifact, Pipeline } from "@aws-cdk/aws-codepipeline";
import {
  CloudFormationCreateUpdateStackAction,
  CodeBuildAction,
  GitHubSourceAction,
} from "@aws-cdk/aws-codepipeline-actions";
import { SecretValue } from "@aws-cdk/core";
import {
  BuildSpec,
  LinuxBuildImage,
  PipelineProject,
} from "@aws-cdk/aws-codebuild";

export class PipelineStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const pipeline = new Pipeline(this, "Pipeline", {
      pipelineName: "Pipeline",
      crossAccountKeys: false,
      restartExecutionOnUpdate: true,
    });

    const cdkSourceOutput = new Artifact("CDKSourceOutput");
    const serviceSourceOutput = new Artifact("ServiceSourceOutput");

    pipeline.addStage({
      stageName: "Source",
      actions: [
        new GitHubSourceAction({
          owner: "Gtofig",
          repo: "aws-pipeline",
          branch: "master",
          actionName: "Pipeline_Source",
          oauthToken: SecretValue.secretsManager("github-token"),
          output: cdkSourceOutput,
        }),
        new GitHubSourceAction({
          owner: "Gtofig",
          repo: "express-lambda",
          branch: "master",
          actionName: "Service_Source",
          oauthToken: SecretValue.secretsManager("github-token"),
          output: serviceSourceOutput,
        }),
      ],
    });

    const cdkBuildOutput = new Artifact("CdkBuildOutput");
    const serviceBuildOutput = new Artifact("ServiceBuildOutput");

    pipeline.addStage({
      stageName: "Build",
      actions: [
        new CodeBuildAction({
          actionName: "CDK_Build",
          input: cdkSourceOutput,
          outputs: [cdkBuildOutput],
          project: new PipelineProject(this, "CdkBuildProject", {
            environment: {
              buildImage: LinuxBuildImage.STANDARD_5_0,
            },
            buildSpec: BuildSpec.fromSourceFilename(
              "build-specs/cdk-build-spec.yml"
            ),
          }),
        }),
        new CodeBuildAction({
          actionName: "Service_Build",
          input: serviceSourceOutput,
          outputs: [serviceBuildOutput],
          project: new PipelineProject(this, "ServiceBuildProject", {
            environment: {
              buildImage: LinuxBuildImage.STANDARD_5_0,
            },
            buildSpec: BuildSpec.fromSourceFilename(
              "build-specs/service-build-spec.yml"
            ),
          }),
        }),
      ],
    });

    pipeline.addStage({
      stageName: "Pipeline_Update",
      actions: [
        new CloudFormationCreateUpdateStackAction({
          actionName: "Pipeline_Update",
          stackName: "PipelineStack",
          templatePath: cdkBuildOutput.atPath("PipelineStack.template.json"),
          adminPermissions: true,
        }),
      ],
    });
  }
}
