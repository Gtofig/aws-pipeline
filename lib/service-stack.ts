import { Construct, Stack, StackProps } from "@aws-cdk/core";
import { Code, Function, Runtime } from "@aws-cdk/aws-lambda";
import { HttpApi } from "@aws-cdk/aws-apigatewayv2";
import { LambdaProxyIntegration } from "@aws-cdk/aws-apigatewayv2-integrations";

export class ServiceStack extends Stack {
  public readonly serviceCode: Code;
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    this.serviceCode = Code.fromCfnParameters();

    const lambda = new Function(this, "ServiceLambda", {
      runtime: Runtime.NODEJS_14_X,
      handler: "src/lambda.handler",
      code: this.serviceCode,
      functionName: "ServiceLambda",
    });

    new HttpApi(this, "ServiceAPI", {
      defaultIntegration: new LambdaProxyIntegration({
        handler: lambda,
      }),
      apiName: "MyService",
    });
  }
}
