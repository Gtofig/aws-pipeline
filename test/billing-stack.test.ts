import { App } from "@aws-cdk/core";
import { BillingStack } from "../lib/billing-stack";
import { SynthUtils } from "@aws-cdk/assert";

test("Billing Stack", () => {
  const app = new App();
  const stack = new BillingStack(app, "BillingStack", {
    budgetAmount: 1,
    emailAddress: "test@example.com",
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
