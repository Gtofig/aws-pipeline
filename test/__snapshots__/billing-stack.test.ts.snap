// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Billing Stack 1`] = `
Object {
  "Resources": Object {
    "BudgetED13AE41": Object {
      "Properties": Object {
        "Budget": Object {
          "BudgetLimit": Object {
            "Amount": 1,
            "Unit": "USD",
          },
          "BudgetName": "Monthly Budget",
          "BudgetType": "COST",
          "TimeUnit": "MONTHLY",
        },
        "NotificationsWithSubscribers": Array [
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 100,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": "test@example.com",
                "SubscriptionType": "EMAIL",
              },
            ],
          },
        ],
      },
      "Type": "AWS::Budgets::Budget",
    },
  },
}
`;
